File: Dynamic Programming/Cpp/2156.cpp
// 포도주 시식
#include <iostream>
#include <vector>
using namespace std;
int max(int a, int b) {
    return a > b ? a : b;
}
int main() {
    int n;
    cin >> n;
    vector<int> wine(n + 1, 0);
    vector<int> d(n + 1, 0);
    for (int i = 1; i <= n; i++)
        cin >> wine[i];
    d[1] = wine[1];
    d[2] = wine[1] + wine[2];
    for (int i = 3; i <= n; i++) {
        d[i] = max(
            max(d[i - 2] + wine[i], d[i - 3] + wine[i - 1] + wine[i]),
            d[i - 1]);
    }
    cout << d[n];
}

File: Dynamic Programming/Python/2156.py
n = int(input())
wine = [0]
wine.extend([int(input()) for _ in range(n)])
wine.extend([0, 0, 0, 0])
d = [0, wine[1], wine[2] + wine[1]]
for i in range(3, n+1):
    d.append(max(
        d[i - 2] + wine[i],
        d[i - 3] + wine[i - 1] + wine[i],
        d[i - 1]))
print(d[n])
